pkg/distributor/query.go

```diff
@@ -433,8 +433,12 @@ func (r *ingesterQueryResult) receiveResponse(stream ingester_client.Ingester_Qu
 		r.chunkseriesBatches = append(r.chunkseriesBatches, resp.Chunkseries)
 	} else if len(resp.StreamingSeries) > 0 {
 		labelsBatch := make([]labels.Labels, 0, len(resp.StreamingSeries))
+		var (
+			labelsBuilder labels.ScratchBuilder
+			l             labels.Labels
+		)
 		for _, s := range resp.StreamingSeries {
-			l := mimirpb.FromLabelAdaptersToLabelsWithCopy(s.Labels)
+			mimirpb.FromLabelAdaptersOverwriteLabelsSafe(&labelsBuilder, s.Labels, &l)
```

Comment:
It seems like this is going to override `l labels.Labels` on every iteration leading to corrupted results?

Reply:
I see that a copy is made to fix an issue in a later commit. I'm struggling to understand if this is safe or not. Maybe it's easier to return the newly created `label.Labels` if we're making copies anyway?

Reply:
The variable `l labels.Labels` is copied when appending to `labelsBatch`:

```go
labelsBatch = append(labelsBatch, l)
```

> I see that a copy is made to fix an issue in a later commit

This fix is specific to `stringlabels` build mode, where I realized I had to copy the `ScratchBuilder`'s internal buffer.